javascripts/2048.js

```diff
@@ -1,32 +1,400 @@
+    createTile();
+    createTile();
+
+    score = 0; // sets score to 0 for a new game
+    alreadyWon = false;
+    gameOver = false;
+    changeDisplayedScore();
+  }
+
+  $('body').keydown(function(event) {
+    if (!gameOver) {
+      var arrowKeys = [37, 38, 39, 40];
+      if (arrowKeys.indexOf(event.which) > -1) {
+        moveTiles(event.which);
+        matched(event.which);
+        moveTiles(event.which);
+        if (!isBoardFull() && actionOccurred) {
+          createTile();
+        }
+
+        if (!alreadyWon && hasWon()) {
+          $('img').addClass("blastoff").on("animationend", function() { $(this).removeClass("blastoff") });
+          setTimeout (function(){alert("Congratulations, YOU HAVE WOOOOOON!!!")}, 3000);
+          alreadyWon = true;
+        } else if (alreadyWon && hasLost()) {
+          alert("Congrats on winning!\nBut there are no more moves for you to make.\nPlease start a new game.")
+          gameOver = true;
+        } else if (hasLost()) {
+          $('img').addClass("crashandburn").on("animationend", function() { $(this).removeClass("crashandburn") });
+          setTimeout (function(){alert("YOU HAVE FAILED! D:")}, 1000);
+          gameOver = true;
+        }
+        actionOccurred = false;
+        event.preventDefault();
+      }
     }
   })
-})
+});
 
-function moveTile(tile, direction) {
-  var new_tile_value = tile.attr("data-val") * 2;
-  tile.attr("data-val", new_tile_value);
-  tile.text(new_tile_value);
+// tile creation ------------------------------------------------------
+function randomizeValue() {
+  var fourSidedDie = Math.ceil(Math.random() * 4 - 1) + 1
+
+  if (fourSidedDie == 4) {
+    return 4;
+  } else {
+    return 2;
+  }
+}
 
+function randomizeLocation() {
+  // floor rounds down for an integer
+  var row = Math.floor(
+    Math.random() *(BOARDSIZE - MINBOARDLOCALE) + MINBOARDLOCALE );
+  var col = Math.floor(
+    Math.random() *(BOARDSIZE - MINBOARDLOCALE) + MINBOARDLOCALE );
+
+  // need to check if slot is empty
+  while (!empty(board[row][col])) {  // can probably refactor this
+    var row = Math.floor(
+      Math.random() *(BOARDSIZE - MINBOARDLOCALE) + MINBOARDLOCALE );
+    var col = Math.floor(
+      Math.random() *(BOARDSIZE - MINBOARDLOCALE) + MINBOARDLOCALE );
+    // need to check if slot is empty
+  }
+  return [row, col];
+}
+
+function createTile() {
+  var location = randomizeLocation()
+  var row = location[0];
+  var col = location[1];
+  var value = randomizeValue();
+  board[row][col] = value;
+  createVisualTile(row, col, value);
+}
+
+// matching & empty checking ---------------------------------------------------
+function empty(location) {
+  // input will be board location
+  // check if board array location is undefined
+  var answer = (location == undefined) ? true : false;
+  return answer;
+}
+
+function matched(direction) {
   switch(direction) {
     case 38: //up
-      tile.attr("data-row","r0");
+      var rowStart = 0; //rowstart
```

Comment:
I'm sure it bothered you to have this repeated code. What are your options for factoring it into its own, reusable function? What are the bits that change? In order to traverse the board, you're adding a particular value to `r` and another to `c`, right? A value in `[-1, 0, 1]`?